lib/active_model/serializer/adapter/json_api/deserialization.rb

```diff
@@ -0,0 +1,115 @@
+module ActiveModel
+  class Serializer
+    module Adapter
+      class JsonApi
+        module Deserialization
+          module_function
+
+          # Transform a JSON API document, containing a single data object, from a hash
+          # into an ActiveRecord#attributes hash.
+          #
+          # @param [Hash/ActionController::Parameters] document
+          # @param [Hash] options
+          #   fields: Array of symbols and a Hash. Specify whitelisted fields, optionally
+          #     specifying the attribute name on the model.
+          # @return [Hash] ActiveRecord-ready hash
+          #
+          # @example
+          #   document = {
+          #     data: {
+          #       id: 1,
+          #       type: 'post',
+          #       attributes: {
+          #         title: 'Title 1',
+          #         date: '2015-12-20'
+          #       },
+          #       associations: {
+          #         author: {
+          #           type: 'user',
+          #           id: 2
+          #         },
+          #         comments: [{
+          #           type: 'comment',
+          #           id: 3
+          #         },{
+          #           type: 'comment',
+          #           id: 4
+          #         }]
+          #       }
+          #     }
+          #   }
+          #
+          #   parse(document) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   date: '2015-12-20',
+          #     #   author_id: 2,
+          #     #   comment_ids: [3, 4]
+          #     # }
+          #
+          #   parse(document, fields: [:title, :author, date: :published_at]) #=>
+          #     # {
+          #     #   title: 'Title 1',
+          #     #   published_at: '2015-12-20',
+          #     #   author_id: 2,
+          #     # }
+          #
+          def parse(document, options = {})
+            fields = parse_fields_option(options[:fields])
+
+            document = document.dup.permit!.to_h if document.is_a?(ActionController::Parameters)
+
+            hash = {}
+
+            primary_data = document.fetch('data', {})
+            hash[:id] = primary_data['id'] if primary_data['id'] && (fields.nil? || fields[:id])
+
+            hash.merge!(parse_attributes(primary_data['attributes'], fields))
+            hash.merge!(parse_relationships(primary_data['relationships'], fields))
+
+            hash
+          end
+
+          # @api private
+          def parse_fields_option(fields)
+            return nil unless fields.is_a?(Array)
+            fields.each_with_object({}) do |attr, hash|
+              if attr.is_a?(Symbol)
+                hash[attr] = attr
+              elsif attr.is_a?(Hash)
+                hash.merge!(attr)
+              end
+            end
+          end
+
+          # @api private
+          def parse_attributes(attributes, fields)
```

Comment:
?

1. Given fields and truthy attributes
  1. Build a hash from each key/value pair
    1. If fields is truthy
      1. And there is a value for the symbolized key
         1. Add that value as the key for the original hash's value
     1.  And the key can be symbolized
        1. Add that symbolized key as the key for the original hash's value
1. Else `{}`